### PR TITLE
[extractor/eplus] Add extractor for URLs from "ib.eplus.jp"

### PR DESCRIPTION
**Pull request by pzhlkj6612**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

</details>

### Commit message

For VoD and live on the [inbound site of "eplus"](https://ib.eplus.jp/) (Japanese: イープラス).

Only works for the link starts with "live.eplus.jp/ex/player?ib=".

Using this extractor will count as one "device" for the current event.

Thanks to **Peter Rowlands** ([pmrowla](https://github.com/pmrowla)) and their "[pmrowla/streamlink-plugins](https://github.com/pmrowla/streamlink-plugins)".

### The generated entry in "supportedsites.md"

- **eplus:inbound**: e+ (イープラス). Only works for the link starts with "live.eplus.jp/ex/player?ib="

### Why I mentioned "ib.eplus.jp"

"eplus.jp" serves users in Japan, while "ib.eplus.jp" provides services for overseas users. I'm not living in Japan, so there is no way to buy tickets with an eplus account. Without a ticket, I'm not able to know the form of URLs and the content of webpages.

### Why only-matching tests?

All usable URLs are from purchased events. Posting a real URL might give eplus a chance to find users who shared their tickets.

### Question

How to update HTTP request headers for "HlsFD" during download? Cookies expire after one hour. We need to visit the passed-in URL and replace old cookies with the new ones in the response headers.

#### The structure of "data_json" ("app" in the webpage)

<details>
<summary>JSON object</summary>

```json
{
    "app_id": "...",
    "app_name": "...",
    "venue_name": "\uff33\uff54\uff52\uff45\uff41\uff4d\uff49\uff4e\uff47\uff0b", // "Ｓｔｒｅａｍｉｎｇ＋"
    "stream_type": "AWS_SINGLE",
    "embedded_type": "IFRAME",
    "player_mode": null,
    "delivery_status": "...",
    "stream_status": "...",
    "number_of_channel": "1",
    "event_datetime": "...",
    "event_datetime_text": "...",
    "buy_ticket_url": "...",
    "description": null,
    "page_display_period": "...",
    "page_display_period_text": "...",
    "is_live_stream": true,
    "chat_mode": "OFF",
    "drm_mode": null,
    "drm_strict_mode": null,
    "virtual_live_mode": null,
    "chat_type": null,
    "chat_open_mode": "OFF",
    "stream_end_at": null,
    "chat_embed_link": null,
    "archive_mode": "ON",
    "archived_status": "...",
    "image_url": null,
    "end_event_image_url": null,
    "all_in_one_thumbnail": "ON",
    "content": "...",
    "allow_number_of_viewable": "1",
    "event_type": "1",
    "term_event_from": "...",
    "term_event_to": "...",
    "is_pass_ticket": "NO",
    "delete_flag": "0",
    "copyright_code": "...",
    "shooting_angle": "",
    "extra": null,
    "extra_archive": null,
    "is_dvr": "OFF",
    "archive_time": "...",
    "archive_kubun": "1",
    "zoom_date_text": {
        "year": "...",
        "month_day_from": "...",
        "month_day_to": "...",
        "day_of_week_from": "...",
        "day_of_week_to": "...",
        "time": "..."
    },
    // The following 4 fields will disappear if the event becomes inaccessible.
    "emergency_message": null,
    "emergency_mode": "OFF",
    "emergency_image": "https:\/\/d1mc0giw38g3ps.cloudfront.net\/stream_images\/emergency.jpg",
    "origin_chat_embed_link": null
}

```

</details>

#### The lifecycle of an event

<table>
<thead>
  <tr>
    <th colspan="2">What I saw<br></th>
    <th rowspan="2">Timeline ↓</th>
    <th colspan="4">In "data_json" ("app" in the webpage)</th>
    <th>In the webpage</th>
  </tr>
  <tr>
    <th>Playable on the webpage?</th>
    <th>Status of the event</th>
    <th>"delivery_status"</th>
    <th>"stream_status"</th>
    <th>"archived_status"</th>
    <th>"is_live_stream"</th>
    <th>m3u8 url (listChannels)</th>
  </tr>
</thead>
<tbody>
  <tr>
    <td rowspan="4">no</td>
    <td rowspan="5">not started</td>
    <td></td>
    <td rowspan="4">PREPARING</td>
    <td>PREPARING</td>
    <td rowspan="7">null</td>
    <td rowspan="3">false</td>
    <td rowspan="2">[]</td>
  </tr>
  <tr>
    <td>T0-900mins</td>
    <td>CREATING</td>
  </tr>
  <tr>
    <td>T0-900mins</td>
    <td>READY</td>
    <td rowspan="6">["https://stream.live.eplus.jp/out/v1/{hash0}/index.m3u8"]</td>
  </tr>
  <tr>
    <td>T0-240mins</td>
    <td rowspan="4">LIVE</td>
    <td rowspan="4">true</td>
  </tr>
  <tr>
    <td rowspan="2">yes</td>
    <td>T0</td>
    <td rowspan="2">STARTED</td>
  </tr>
  <tr>
    <td>started</td>
    <td>T0+60mins</td>
  </tr>
  <tr>
    <td>forgot</td>
    <td rowspan="2">ended</td>
    <td>T1-10mins</td>
    <td>STOPPED</td>
  </tr>
  <tr>
    <td>no</td>
    <td>T1</td>
    <td>WAIT_CONFIRM_ARCHIVED</td>
    <td rowspan="3">READY</td>
    <td rowspan="5">SUCCEEDED</td>
    <td rowspan="3">false</td>
  </tr>
  <tr>
    <td rowspan="3">yes</td>
    <td rowspan="3">archive available</td>
    <td>T1+90mins</td>
    <td rowspan="4">CONFIRMED_ARCHIVE</td>
    <td>["https://vod.live.eplus.jp/out/v1/{hash0}/{hash1}/{name0}.m3u8"]</td>
  </tr>
  <tr>
    <td>T1+180mins</td>
    <td rowspan="2">["https://vod.live.eplus.jp/out/v1/{hash0}/{name1}/{name1}.m3u8"]</td>
  </tr>
  <tr>
    <td>T1+980mins</td>
    <td rowspan="2">DELETED</td>
    <td rowspan="2">true</td>
  </tr>
  <tr>
    <td>no</td>
    <td>expired</td>
    <td></td>
    <td>non-existing</td>
  </tr>
</tbody>
</table>

The above table is generated by [HTML Tables generator – TablesGenerator.com](https://www.tablesgenerator.com/html_tables).

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

**Comment by pukkandan** [comment]
> How to update HTTP request headers for "HlsFD" during download? Cookies expire after one hour. We need to visit the passed-in URL and replace old cookies with the new ones in the response headers.

We currently don't have any mechanism to do this. One possible way is to create a custom downloader, but considering it only helps for archives, it is likely not worth it (is it likely archive downloads to take over an hour?)

> All usable URLs are from purchased events. Posting a real URL might give eplus a chance to find users who shared their tickets.

Fair enough, Just add and run a test locally - no need to push it

**Comment by pukkandan** [comment]
Pls avoid force-pushing in future. Dirty commits are fine

**Comment by pzhlkj6612** [comment]
> Pls avoid force-pushing in future. Dirty commits are fine

Oh sorry, I won't do that.

File: yt_dlp/extractor/eplus.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        'url': 'https://live.eplus.jp/ex/player?ib=YA4Z%2Fz2rpI5cDl3V%2Bx9PI%2FXTGOX0j8IBzPthgx7%2BieWeff6iLcdsdsds1926zosZU9AbB3gdL3wG%2BNa1afQdIf%3D%3D',
+        'only_matching': True,
+    }]
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+
+        data_json = self._parse_json(self._html_search_regex(
+            r'<script>\s*var app = (.+);\n', webpage, 'data_json'), video_id)
```
Comment (pukkandan) [inline yt_dlp/extractor/eplus.py:31]: self._search_json

File: yt_dlp/extractor/eplus.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+        delivery_status = traverse_obj(data_json, 'delivery_status')
+        archive_mode = traverse_obj(data_json, 'archive_mode')
+
+        self.to_screen(f'delivery_status = {delivery_status}, archive_mode = {archive_mode}')
```
Comment (pukkandan) [inline yt_dlp/extractor/eplus.py:36]: Remove, or write_debug

File: yt_dlp/extractor/eplus.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        elif delivery_status == 'WAIT_CONFIRM_ARCHIVED':
+            raise UserNotLive('The archive will be available shortly.')
+        elif delivery_status == 'CONFIRMED_ARCHIVE':
+            # The archive may be available.
+            # TODO: how to update HTTP request headers for "HlsFD" during download?
+            self.report_warning(
+                'Due to technical limitations, the download will be interrupted after one hour. You can restart to continue the download.')
+            live_status = 'was_live'
+        else:
+            raise ExtractorError(f'Unknown delivery_status: {delivery_status}', expected=False)
```
Comment (pukkandan) [inline yt_dlp/extractor/eplus.py:61]:
```suggestion
        if delivery_status == 'PREPARING':
            raise UserNotLive('This event has not started yet')
        elif delivery_status == 'STARTED':
            # FIXME: HTTP request headers need to be updated to continue download
            self.report_warning(
                'Due to technical limitations, the download will be interrupted after one hour')
            live_status = 'is_live'
        elif delivery_status == 'STOPPED':
            if archive_mode == 'ON':  # I didn't see values other than "ON"
                raise UserNotLive(
                    'This event has ended, but the archive has not been generated yet')
            raise ExtractorError(
                'This event has ended and there is no archive for this event', expected=True)
        elif delivery_status == 'WAIT_CONFIRM_ARCHIVED':
            raise UserNotLive(
                'This event has ended, and the archive will be available shortly')
        elif delivery_status == 'CONFIRMED_ARCHIVE':
            # FIXME: HTTP request headers need to be updated to continue download
            self.report_warning(
                'Due to technical limitations, the download will be interrupted after one hour. '
                'You can restart to continue the download')
            live_status = 'was_live'
        else:
            raise ExtractorError(f'Unknown delivery_status: {delivery_status}')
```

File: yt_dlp/extractor/eplus.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+        # TODO: if there are multiple playlists, tell me.
```
Comment (pukkandan) [inline yt_dlp/extractor/eplus.py:63]:
```suggestion
```

File: yt_dlp/extractor/eplus.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        m3u8_playlist_urls = self._html_search_regex(
+            r'var listChannels = (.+);\n', webpage, 'listChannels', default=None)
```
Comment (pukkandan) [inline yt_dlp/extractor/eplus.py:65]: _search_json

File: yt_dlp/extractor/eplus.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+        if not m3u8_playlist_urls:
+            raise ExtractorError(
+                'Could not find the playlist URL. This event may not be accessible.', expected=True)
```
Comment (pukkandan) [inline yt_dlp/extractor/eplus.py:68]:
```suggestion
        if not m3u8_playlist_urls:
            self.raise_no_formats(
                'Could not find the playlist URL. This event may not be accessible', expected=True)
```

File: yt_dlp/extractor/eplus.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+                m3u8_playlist_urls, video_id), 0), video_id=video_id),
+
+            # optional metafields
+
+            'live_status': live_status,
+            'description': traverse_obj(data_json, 'content'),
+            'timestamp': calendar.timegm((datetime.datetime.strptime(
+                traverse_obj(data_json, 'event_datetime'), '%Y%m%d%H%M')
+                - datetime.timedelta(hours=9)).timetuple()),
+        }
```
Comment (pukkandan) [inline yt_dlp/extractor/eplus.py:85]:
```suggestion
        return {
            'id': data_json['app_id'],
            'title': data_json.get('app_name'),
            'formats': self._extract_m3u8_formats(m3u8_playlist_url, video_id),
            'live_status': live_status,
            'description': traverse_obj(data_json, 'content'),
            'timestamp': try_call(lambda: unified_timestamp(data_json['event_datetime']) - 32400),
        }
```

File: yt_dlp/extractor/eplus.py [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+            'formats': self._extract_m3u8_formats(m3u8_url=traverse_obj(self._parse_json(
+                m3u8_playlist_urls, video_id), 0), video_id=video_id),
+
+            # optional metafields
+
+            'live_status': live_status,
+            'description': traverse_obj(data_json, 'content'),
+            'timestamp': calendar.timegm((datetime.datetime.strptime(
+                traverse_obj(data_json, 'event_datetime'), '%Y%m%d%H%M')
+                - datetime.timedelta(hours=9)).timetuple()),
```
Comment (pukkandan) [inline yt_dlp/extractor/eplus.py:84]:
```suggestion
            'timestamp': try_call(lambda: unified_timestamp(data_json['event_datetime']) - 32400),
```

**Review by pukkandan — CHANGES_REQUESTED** [review]
Note that `UserNotLive` must be raised only if you the URL **may go live** some time in the future. If the URL will never become live again, `self.raise_no_formats` instead. This convention is needed for `--wait-for-video` and `--ignore-no-formats` to work correctly

File: yt_dlp/extractor/eplus.py [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+from ..utils import (
+    ExtractorError,
+    UserNotLive,
+    traverse_obj,
+)
+
+
+class EplusIbIE(InfoExtractor):
+    IE_NAME = 'eplus:inbound'
+    IE_DESC = 'e+ (イープラス). Only works for the link starts with "live.eplus.jp/ex/player?ib="'
```
Comment (pukkandan) [inline yt_dlp/extractor/eplus.py:14]:
```suggestion
    IE_DESC = 'e+ (イープラス)'
```

File: yt_dlp/extractor/eplus.py [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+        if delivery_status == 'PREPARING':
+            raise UserNotLive('This event has not started yet')
+        elif delivery_status == 'STARTED':
+            # FIXME: HTTP request headers need to be updated to continue download
+            self.report_warning(
+                'Due to technical limitations, the download will be interrupted after one hour')
+            live_status = 'is_live'
+        elif delivery_status == 'STOPPED':
+            if archive_mode == 'ON':  # I didn't see values other than "ON"
```
Comment (pukkandan) [inline yt_dlp/extractor/eplus.py:46]:
```suggestion
            if archive_mode == 'ON':
```

File: yt_dlp/extractor/eplus.py [hunk 11]
```diff
@@ -11,7 +11,7 @@
 
 class EplusIbIE(InfoExtractor):
     IE_NAME = 'eplus:inbound'
-    IE_DESC = 'e+ (イープラス). Only works for the link starts with "live.eplus.jp/ex/player?ib="'
+    IE_DESC = 'e+ (イープラス)'
```
Comment (pzhlkj6612) [inline yt_dlp/extractor/eplus.py:13]: Well, the eplus player links for local and overseas users may be different. I'm not able to buy tickets as a local, so this extractor, now, is only for extracting info from inbound (ib) links.

Local users will not use this extractor if they see that description, so I think it's better to keep the previous text here.

```suggestion
    IE_DESC = 'e+ (イープラス). Only works for the link starts with "live.eplus.jp/ex/player?ib="'
```

**Review by pzhlkj6612 — COMMENTED** [review]
Hello, @pukkandan! Thanks for your detailed review. I've applied some fixes for your suggestions, and the code now looks cleaner.

However, I'm afraid I can't agree with one change you made. Here is my review.

File: yt_dlp/extractor/eplus.py (hunk 11 above)
Comment (pukkandan) [reply yt_dlp/extractor/eplus.py:13]: If the URL does not match, yt-dlp will give "Unsupported URL" error. It's not IE_DESC's job to write out the matching URL patterns. Maybe you'd like to change the description to `e+ overseas` instead

Comment (pzhlkj6612) [reply yt_dlp/extractor/eplus.py]: Got it, and thanks for your nice suggestion. See 157ee41b6c03364d6ceafbb76c13d2b1767cabfb .

**Comment by pzhlkj6612** [comment]
After applying 157ee41b6c03364d6ceafbb76c13d2b1767cabfb , the generated entry in "supportedsites.md" has changed to:

- **eplus:inbound**: e+ (イープラス) overseas

File: yt_dlp/extractor/eplus.py [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+            # FIXME: HTTP request headers need to be updated to continue download
+            self.report_warning(
+                'Due to technical limitations, the download will be interrupted after one hour. '
+                'You can restart to continue the download')
+            live_status = 'was_live'
+        else:
+            raise ExtractorError(f'Unknown delivery_status: {delivery_status}')
+
+        m3u8_playlist_urls = self._search_json(
+            r'var listChannels = ', webpage, 'listChannels', video_id, contains_pattern=r'\[.+\]', default=[])
```
Comment (pukkandan) [inline yt_dlp/extractor/eplus.py:62]:
```suggestion
            r'var listChannels\s*=', webpage, 'hls URLs', video_id, contains_pattern=r'\[.+\]', default=[])
```

File: yt_dlp/extractor/eplus.py [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+        # complex fake url
+        'url': 'https://live.eplus.jp/ex/player?ib=YA4Z%2Fz2rpI5cDl3V%2Bx9PI%2FXTGOX0j8IBzPthgx7%2BieWeff6iLcdsdsds1926zosZU9AbB3gdL3wG%2BNa1afQdIf%3D%3D',
+        'only_matching': True,
+    }]
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+
+        data_json = self._search_json(r'<script>\s*var app = ', webpage, 'data_json', video_id)
```
Comment (pukkandan) [inline yt_dlp/extractor/eplus.py:29]:
```suggestion
        data_json = self._search_json(r'<script>\s*var app\s*=', webpage, 'data json', video_id)
```

**Review by pukkandan — COMMENTED** [review]
Why's this drafted?

**Comment by pzhlkj6612** [comment]
@pukkandan ,

> Note that `UserNotLive` must be raised only if you the URL **may go live** some time in the future. If the URL will never become live again, `self.raise_no_formats` instead. This convention is needed for `--wait-for-video` and `--ignore-no-formats` to work correctly

Yes, I understand.

> Why's this drafted?

Now this PR is waiting for review.

File: yt_dlp/extractor/eplus.py [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+    ExtractorError,
+    try_call,
+    unified_timestamp,
+)
+
+
+class EplusIbIE(InfoExtractor):
+    IE_NAME = 'eplus:inbound'
+    IE_DESC = 'e+ (イープラス) overseas'
+    _VALID_URL = r'https?://live\.eplus\.jp/ex/player\?ib=(?P<id>([A-Za-z\d]|%2B|%2F){86}(%3D){2})'
```
Comment (pukkandan) [inline yt_dlp/extractor/eplus.py:12]:
```suggestion
    _VALID_URL = r'https?://live\.eplus\.jp/ex/player\?ib=(?P<id>(?:\w|%2B|%2F){86}%3D%3D)'
```

File: yt_dlp/extractor/eplus.py [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+    _TESTS = [{
```
Comment (pukkandan) [inline yt_dlp/extractor/eplus.py:13]: Why are all tests "fake"? Use a real test that currently works for you locally and then add `'skip': 'live'`

File: yt_dlp/extractor/eplus.py [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+            if archive_mode != 'ON':
+                raise ExtractorError(
+                    'This event has ended and there is no archive for this event', expected=True)
+            live_status = 'is_upcoming'
+        elif delivery_status == 'WAIT_CONFIRM_ARCHIVED':
+            live_status = 'is_upcoming'
+        elif delivery_status == 'CONFIRMED_ARCHIVE':
+            live_status = 'was_live'
+        else:
+            raise ExtractorError(f'Unknown delivery_status: {delivery_status}')
```
Comment (pukkandan) [inline yt_dlp/extractor/eplus.py:72]: Don't raise, issue warning try to extract as-if `is_live`.

File: yt_dlp/extractor/eplus.py [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+        elif live_status == 'is_upcoming':
+            if delivery_status == 'PREPARING':
+                self.raise_no_formats(f'This live event will begin at {release_timestamp_str} JST', expected=True)
+            else:
+                self.raise_no_formats('This event has ended, and the archive will be available shortly', expected=True)
+        else:
+            for m3u8_playlist_url in m3u8_playlist_urls:
+                formats.extend(self._extract_m3u8_formats(m3u8_playlist_url, video_id))
+
+        if formats:
```
Comment (pukkandan) [inline yt_dlp/extractor/eplus.py:94]:
```suggestion
```

File: yt_dlp/extractor/eplus.py [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+                self.report_warning(f'{warning}. You can restart to continue the download')
+
+        return {
+            'id': data_json['app_id'],
+            'title': data_json.get('app_name'),
+            'formats': formats,
+            'live_status': live_status,
+            'description': data_json.get('content'),
+            'timestamp': release_timestamp,
+            'release_timestamp': release_timestamp,
```
Comment (pukkandan) [inline yt_dlp/extractor/eplus.py:109]: Why both?

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/eplus.py (hunk 18 above)
Comment (pzhlkj6612) [reply yt_dlp/extractor/eplus.py]: I thought "`timestamp`" was used for something else than "`release_timestamp`", so kept it in "info_dict".

File: yt_dlp/extractor/eplus.py (hunk 15 above)
Comment (pzhlkj6612) [reply yt_dlp/extractor/eplus.py:13]: Well, the reason is:

> All usable URLs are from purchased events. Posting a real URL might give eplus a chance to find users who shared their tickets.

File: yt_dlp/extractor/eplus.py (hunk 14 above)
Comment (pzhlkj6612) [reply yt_dlp/extractor/eplus.py]: Underscores are also matched, but it doesn't hurt too much.

File: yt_dlp/extractor/eplus.py (hunk 15 above)
Comment (pukkandan) [reply yt_dlp/extractor/eplus.py:13]: Sorry, forgot about that

File: yt_dlp/extractor/eplus.py [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+            live_status = 'is_upcoming'
+        elif delivery_status == 'STARTED':
+            live_status = 'is_live'
+        elif delivery_status == 'STOPPED':
+            if archive_mode != 'ON':
+                raise ExtractorError(
+                    'This event has ended and there is no archive for this event', expected=True)
+            live_status = 'is_upcoming'
+        elif delivery_status == 'WAIT_CONFIRM_ARCHIVED':
+            live_status = 'is_upcoming'
```
Comment (bashonly) [inline yt_dlp/extractor/eplus.py:59]: Do we want `--wait-for-video` to wait for the VOD? Seems like `post_live` might be more appropriate. Although if we can be sure `--wait-for-video` will not be endless, then it is fine as-is
```suggestion
        elif delivery_status == 'WAIT_CONFIRM_ARCHIVED':
            live_status = 'post_live'
```

File: yt_dlp/extractor/eplus.py [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+                self.raise_no_formats(
+                    f'Could not find the playlist URL. This live event will begin at {release_timestamp_str} JST', expected=True)
+            else:
+                self.raise_no_formats(
+                    'Could not find the playlist URL. This event may not be accessible', expected=True)
+        elif live_status == 'is_upcoming':
+            if delivery_status == 'PREPARING':
+                self.raise_no_formats(f'This live event will begin at {release_timestamp_str} JST', expected=True)
+            else:
+                self.raise_no_formats('This event has ended, and the archive will be available shortly', expected=True)
```
Comment (bashonly) [inline yt_dlp/extractor/eplus.py:81]: If the above suggestion is applied, you can also do this
```suggestion
        elif live_status == 'is_upcoming':
            self.raise_no_formats(f'This live event will begin at {release_timestamp_str} JST', expected=True)
        elif live_status == 'post_live':
            self.raise_no_formats('This event has ended, and the archive will be available shortly', expected=True)
```

**Review by bashonly — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/eplus.py (hunk 19 above)
Comment (pzhlkj6612) [reply yt_dlp/extractor/eplus.py:57]: Thanks for your review.

<br>

> Do we want `--wait-for-video` to wait for the VOD?

Yes, I think we should wait for the VOD.

> Seems like `post_live` might be more appropriate. Although if we can be sure `--wait-for-video` will not be endless, then it is fine as-is

Hmm, am I incorrectly using the "is_upcoming" live status?

<br>

According to the table in my first comment (https://github.com/yt-dlp/yt-dlp/pull/5782#issue-1491140392), the "WAIT_CONFIRM_ARCHIVED" status will eventually switch to "CONFIRMED_ARCHIVE", so the wait won't be infinite.

Comment (bashonly) [reply yt_dlp/extractor/eplus.py:57]: After looking at the core code, I guess it doesn't make much of a difference behavior-wise. IMO `post_live` would be the expected value for a livestream that has ended but hasn't finished processing (this is how `YoutubeIE` does it). `--wait-for-video` will still work with `post_live`

Comment (pzhlkj6612) [reply yt_dlp/extractor/eplus.py]: Oh, I think I've missed something. We have

```
    live_status:    None (=unknown), 'is_live', 'is_upcoming', 'was_live', 'not_live',
                    or 'post_live' (was live, but VOD is not yet processed)              <---
                    If absent, automatically set from is_live, was_live
```

So, to make the extraction information more accurate, the "live_status" should be "post_live".

@bashonly , thank you for pointing me out.

**Pull request merged**